miniwin/miniwin/src/miniwin_d3drmmesh.cpp

```diff
@@ -38,85 +62,158 @@ HRESULT Direct3DRMMeshImpl::GetGroup(
 	return DD_OK;
 }
 
 HRESULT Direct3DRMMeshImpl::SetGroupTexture(int groupIndex, IDirect3DRMTexture* texture)
 {
-	MINIWIN_NOT_IMPLEMENTED();
-	m_groupTexture = texture;
+	m_groups[groupIndex].texture = texture;
 	return DD_OK;
 }
 
 HRESULT Direct3DRMMeshImpl::GetGroupTexture(int groupIndex, LPDIRECT3DRMTEXTURE* texture)
 {
-	MINIWIN_NOT_IMPLEMENTED();
-	if (!m_groupTexture) {
+	auto it = m_groups.find(groupIndex);
+	if (it == m_groups.end() || !it->second.texture) {
 		return DDERR_GENERIC;
 	}
 
-	m_groupTexture->AddRef();
-	*texture = m_groupTexture;
+	it->second.texture->AddRef();
+	*texture = it->second.texture;
+	return DD_OK;
+}
+
+HRESULT Direct3DRMMeshImpl::SetGroupMapping(D3DRMGROUPINDEX groupIndex, D3DRMMAPPING mapping)
+{
 	return DD_OK;
 }
 
 D3DRMMAPPING Direct3DRMMeshImpl::GetGroupMapping(int groupIndex)
 {
-	MINIWIN_NOT_IMPLEMENTED();
 	return D3DRMMAP_PERSPCORRECT;
 }
 
+HRESULT Direct3DRMMeshImpl::SetGroupQuality(int groupIndex, D3DRMRENDERQUALITY quality)
+{
+	m_groups[groupIndex].quality = quality;
+	return DD_OK;
+}
+
 D3DRMRENDERQUALITY Direct3DRMMeshImpl::GetGroupQuality(int groupIndex)
 {
-	MINIWIN_NOT_IMPLEMENTED();
-	return D3DRMRENDER_GOURAUD;
+	auto it = m_groups.find(groupIndex);
+	return (it != m_groups.end()) ? it->second.quality : D3DRMRENDER_GOURAUD;
 }
 
-HRESULT Direct3DRMMeshImpl::GetGroupColor(D3DRMGROUPINDEX index)
+HRESULT Direct3DRMMeshImpl::SetVertices(int groupIndex, int offset, int count, D3DRMVERTEX* vertices)
 {
-	MINIWIN_NOT_IMPLEMENTED();
+	if (count <= 0 || offset < 0) {
+		return DDERR_INVALIDPARAMS;
+	}
+
+	auto& vertList = m_groups[groupIndex].vertices;
+
+	if (offset + count > static_cast<int>(vertList.size())) {
+		vertList.resize(offset + count);
+	}
+
+	std::copy(vertices, vertices + count, vertList.begin() + offset);
 	return DD_OK;
 }
 
 HRESULT Direct3DRMMeshImpl::GetVertices(int groupIndex, int startIndex, int count, D3DRMVERTEX* vertices)
 {
-	MINIWIN_NOT_IMPLEMENTED();
+	if (count <= 0 || startIndex < 0) {
+		return DDERR_INVALIDPARAMS;
+	}
+
+	auto it = m_groups.find(groupIndex);
+	if (it == m_groups.end()) {
+		return DDERR_GENERIC;
+	}
+
+	const auto& vertList = it->second.vertices;
+
+	if (startIndex + count > static_cast<int>(vertList.size())) {
+		return DDERR_INVALIDPARAMS;
+	}
+
+	std::copy(vertList.begin() + startIndex, vertList.begin() + startIndex + count, vertices);
+	return DD_OK;
+}
+
+HRESULT Direct3DRMMeshImpl::GetBox(D3DRMBOX* box)
+{
+	box->min.x = box->min.y = box->min.z = 0;
+	box->max.x = box->max.y = box->max.z = 0;
+
+	for (const auto& [groupIndex, group] : m_groups) {
+		for (const D3DRMVERTEX& v : group.vertices) {
+			box->min.x = std::min(box->min.x, v.position.x);
+			box->min.y = std::min(box->min.y, v.position.y);
+			box->min.z = std::min(box->min.z, v.position.z);
+			box->max.x = std::max(box->max.x, v.position.x);
+			box->max.y = std::max(box->max.y, v.position.y);
+			box->max.z = std::max(box->max.z, v.position.z);
+		}
+	}
```

Comment:
I'm sure we need to cache this if we ever do bounding box tests.
Let's add a `FIXME` here.

Reply:
From what I saw it's called only from dead code.

Reply:
I think `IDirect3DRMViewport::Pick` might use it to quickly discard objects.